dataset/test/bin_test.cpp

```diff
@@ -0,0 +1,194 @@
+  Variable edges_x_coarse =
+      makeVariable<double>(Dims{Dim::X}, Shape{3}, Values{-2, 1, 2});
+  Variable edges_y_coarse =
+      makeVariable<double>(Dims{Dim::Y}, Shape{3}, Values{-2, -1, 2});
+
+  void expect_near(const DataArrayConstView &a, const DataArrayConstView &b) {
+    // "round" last digits for approximate floating point comparison
+    const auto rounding = 100.0 * units::one * buckets::sum(a);
+    EXPECT_EQ(buckets::sum(a) + rounding, buckets::sum(b) + rounding);
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(InputSize, BinTest,
+                         testing::Values(make_table(0), make_table(1),
+                                         make_table(7), make_table(27),
+                                         make_table(1233)));
+
+TEST_P(BinTest, group) {
+  const auto table = GetParam();
+  const auto binned = bin(table, {}, {groups});
+  EXPECT_EQ(binned.dims(), groups.dims());
+}
+
+TEST_P(BinTest, rebin_coarse_to_fine_1d) {
+  const auto table = GetParam();
+  EXPECT_EQ(bin(table, {edges_x}),
+            bin(bin(table, {edges_x_coarse}), {edges_x}));
+}
+
+TEST_P(BinTest, rebin_fine_to_coarse_1d) {
+  const auto table = GetParam();
+  expect_near(bin(table, {edges_x_coarse}),
+              bin(bin(table, {edges_x}), {edges_x_coarse}));
+}
+
+TEST_P(BinTest, 2d) {
+  const auto table = GetParam();
+  const auto x = bin(table, {edges_x});
+  const auto x_then_y = bin(x, {edges_y});
+  const auto xy = bin(table, {edges_x, edges_y});
+  EXPECT_EQ(xy, x_then_y);
+}
+
+TEST_P(BinTest, rebin_coarse_to_fine_2d) {
+  const auto table = GetParam();
+  const auto xy_coarse = bin(table, {edges_x_coarse, edges_y_coarse});
+  const auto xy = bin(table, {edges_x, edges_y});
+  EXPECT_EQ(bin(xy_coarse, {edges_x, edges_y}), xy);
+}
+
+TEST_P(BinTest, rebin_fine_to_coarse_2d) {
+  const auto table = GetParam();
+  const auto xy_coarse = bin(table, {edges_x_coarse, edges_y_coarse});
+  const auto xy = bin(table, {edges_x, edges_y});
+  expect_near(bin(xy, {edges_x_coarse, edges_y_coarse}), xy_coarse);
+}
+
+TEST_P(BinTest, rebin_coarse_to_fine_2d_inner) {
+  const auto table = GetParam();
+  const auto xy_coarse = bin(table, {edges_x_coarse, edges_y_coarse});
+  const auto xy = bin(table, {edges_x_coarse, edges_y});
+  expect_near(bin(xy_coarse, {edges_y}), xy);
+}
+
+TEST_P(BinTest, rebin_coarse_to_fine_2d_outer) {
+  const auto table = GetParam();
+  auto xy_coarse = bin(table, {edges_x_coarse, edges_y});
+  auto xy = bin(table, {edges_x, edges_y});
+  expect_near(bin(xy_coarse, {edges_x}), xy);
+  // Y is inside X and needs to be handled by `bin`, but coord is not required.
+  xy_coarse.coords().erase(Dim::Y);
+  xy.coords().erase(Dim::Y);
+  expect_near(bin(xy_coarse, {edges_x}), xy);
+}
+
+TEST_P(BinTest, group_and_bin) {
+  const auto table = GetParam();
+  const auto x_group = bin(table, {edges_x}, {groups});
+  const auto group = bin(table, {}, {groups});
+  EXPECT_EQ(bin(group, {edges_x}, {}), x_group);
+}
```

Comment:
I think you mentioned in the PR description that `bin` could better account for the need to broadcast . Where can I see that?

Reply:
What I mention in the PR is a behavior change of `broadcast` (how the target dims are specified/determined). It just made the implementation of `bin` simpler, but has no effect on it otherwise.

Reply:
Ok. I'll take a closer look at that.